text/0000-output-options.md

```diff
@@ -0,0 +1,135 @@
+Value of this option from now on will be referred to as `[out-dir]`.
+
+## `--emit` and `--crate-type`
+
+The two options influence what and how many files the compiler has to write to the filesystem. The
+common case compiler having to write one file only (e.g. `--emit` and `--crate-type` only have one
+value each). In this case the path to output file shall be built using these templates:
+
+* `[out-dir]/[filename][.extension]` if [filename] was [inferred][inferred].
+* `[out-dir]/[filename]` otherwise;
+
+[inferred]: #-o-is-not-specified
+
+Otherwise, when there’s multiple files to output, each output file will be written to path built
+using `[out-dir]/[filename][.extension]` template.
+
+What `[filename]` resolves to is specified in [section about `-o`](#-o).
+
+The `[.extension]` is file extension specific to the emitted target. This is a list of extensions
+for each currently supported target:
+
+* asm – `.s`;
+* llvm-bc – `.bc`;
+* llvm-ir – `.ll`;
+* obj – `.o`;
+* dep-info – `.d`;
+* link – depends on `crate-type`:
+  * bin – no extension or `.exe`;
+  * lib – `.rlib`;
+  * rlib – `.rlib`;
+  * dylib – `.so`, `.dll` or `.dylib`;
+  * staticlib – `.a`.
+
+`link` target should prepend prefix `lib` to `[filename]` for all `crate-type`s except `bin`:
+
+    $ rustc foo.rs -o foo --crate-type=staticlib
+    # Output: libfoo.a
+
+## `-o`
+
+The value of `-o` option, unlike `--out-dir` may be adapted by `rustc` in case it conflicts with
+other options. Ignoring the option is strongly discouraged.
+
+Were the value be ignored or adapted, the warning detailing actions taken should be emitted.
+
+### Interactions with `--out-dir`
+
+In absence of `--out-dir` option, `-o` value may contain directory path components:
+
+    $ rustc foo.rs -o bar/foo # A-OK
+
+For this particular command the compiler behaves as if it had gotten options `--out-dir=bar/` and
+`-o foo`. More generally, the compiler interprets the option by splitting value into
+[`filename`][filename] to be used as `[filename]` and [`dirname`][dirname] to be used as
+`[out-dir]`.
+
+[filename]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#tymethod.filename
+[dirname]: http://doc.rust-lang.org/std/path/trait.GenericPath.html#tymethod.dirname
+
+If both `--out-dir` and `-o` are provided, the `-o` value is adjusted so it only contains a
+`filename`, which is used as `[filename]`:
+
+    $ rustc foo.rs --out-dir=baz -o foo # A-OK
+    $ rustc foo.rs --out-dir=baz -o bar/foo
+    warning: `-o foo` is used instead of `-o bar/foo`, because `--out-dir=baz` is specified
+
+### Interactions with `--emit`
+
+The extension has no special meaning in `-o` option values and becomes a part of `[filename]`
+verbatim:
+
+    $ rustc foo.rs --emit=asm,obj -o foo.bar
+    # Output: foo.bar.s foo.bar.o
+    $ rustc foo.rs --crate-type=staticlib,rlib -o foo.bar
+    # Output: foo.bar.rlib foo.bar.a
```

Comment:
Sorry to be pedantic, but do you mean `libfoo.bar.rlib` and `libfoo.bar.a` here?